File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

long long solution(int cap, int n, vector<int> deliveries, vector<int> pickups) {
    long long answer = 0;
    stack<int> deliveryIdx, pickupIdx;

    for (int i = 0; i < n; ++i) {
        if (deliveries[i] != 0)
            deliveryIdx.push(i + 1);
        if (pickups[i] != 0)
            pickupIdx.push(i + 1);
    }

    while (!deliveryIdx.empty() || !pickupIdx.empty()) {
        if (!deliveryIdx.empty() && pickupIdx.empty())
            answer += deliveryIdx.top() * 2;
        else if (deliveryIdx.empty() && !pickupIdx.empty())
            answer += pickupIdx.top() * 2;
        else if (deliveryIdx.top() <= pickupIdx.top())
            answer += pickupIdx.top() * 2;
        else
            answer += deliveryIdx.top() * 2;

        int curCap = cap;

        while (!deliveryIdx.empty() && curCap != 0) {
            int curIdx = deliveryIdx.top();
            if (deliveries[curIdx - 1] <= curCap) {
                curCap -= deliveries[curIdx - 1];
                deliveryIdx.pop();
            }
            else {
                deliveries[curIdx - 1] -= curCap;
                break;
            }
        }

        curCap = cap;

        while (!pickupIdx.empty() && curCap != 0) {
            int curIdx = pickupIdx.top();
            if (pickups[curIdx - 1] <= curCap) {
                curCap -= pickups[curIdx - 1];
                pickupIdx.pop();
            }
            else {
                pickups[curIdx - 1] -= curCap;
                break;
            }
        }
    }
    return answer;
}

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
import java.util.*;
import java.util.stream.*;

class Solution {
    
    class Data{
        int idx;
        int cnt;
        Data(int idx, int cnt){
            this.idx = idx;
            this.cnt = cnt;
        }
    }
    
    
    
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        
        PriorityQueue<Data> deliveryQ = new PriorityQueue<>(Comparator.comparing((Data d) -> d.idx).reversed());
        PriorityQueue<Data> pickupQ = new PriorityQueue<>(Comparator.comparing((Data d) -> d.idx).reversed());
        for(int i = 0 ; i < n ; i++){
            if(deliveries[i] > 0)
                deliveryQ.offer(new Data(i + 1, deliveries[i]));
            if(pickups[i] > 0)
                pickupQ.offer(new Data(i + 1, pickups[i]));
        }
        
        long answer = 0;

        while(!deliveryQ.isEmpty() || !pickupQ.isEmpty()){
            
            // 큐 갱신하기
            int d = 0;
            int maxd = 0;
            while(!deliveryQ.isEmpty() && d < cap){
                Data cur = deliveryQ.poll();
                maxd = Math.max(maxd, cur.idx);
                d += cur.cnt;
                if(d > cap){
                    deliveryQ.offer(new Data(cur.idx, d - cap));
                    d = cap;
                }
            }
 
            int p = 0;
            while(!pickupQ.isEmpty() && p < cap){
                Data cur = pickupQ.poll();
                maxd = Math.max(maxd, cur.idx);
                p += cur.cnt;
                if(p  > cap){
                    pickupQ.offer(new Data(cur.idx, p - cap));
                    p = cap;
                }
            }            
            
            answer += maxd * 2;
            
        }
        
        return answer;
    }
}